### PR TITLE
Fix Async schema loading so schemas cannot be used until they are completely loaded

### PR DESCRIPTION
**Pull request by ColinKerr**
Fixes issue #5326 by caching promises to load schemas and returning information about partially loaded schemas instead of returning partially loaded schemas.  This ensures partially loaded schemas are not accessed and allows us to test for circular schema references with just schema keys for schemas and their references

### CODE REVIEW

File: core/ecschema-metadata/src/Context.ts [hunk 1]
```diff
@@ -47,23 +48,66 @@ export interface ISchemaItemLocater {
...
+
+  private removeSchemaPromise(schemaKey: Readonly<SchemaKey>) {
+    this._schemaPromises = this._schemaPromises.filter((value: LazyLoadedSchema) => !value.matches(schemaKey, SchemaMatchType.Latest));
+  }
+
+  /**
+   * Returns true if the schema exists in either the schema cache or the promise cache.  SchemaMatchType.Latest used.
+   * @param schemaKey The key to search for.
+   */
+  public schemaExits(schemaKey: Readonly<SchemaKey>): boolean {
```
Comment (pmconne) [inline core/ecschema-metadata/src/Context.ts:80]: `s/Exits/Exists`?

Comment (ColinKerr) [reply core/ecschema-metadata/src/Context.ts]: fixed

File: presentation/common/src/presentation-common/KoqPropertyValueFormatter.ts [hunk 2]
```diff
@@ -59,7 +59,7 @@ interface FormattingProps {
   persistenceUnitName: string;
 }
 
-async function getFormattingProps(schemaLocater: ISchemaLocater, options: FormatOptions): Promise<FormattingProps | undefined> {
```
Comment (ColinKerr) [inline presentation/common/src/presentation-common/KoqPropertyValueFormatter.ts:62]: I think it was a mistake to use ISchemaLocater here ... in all cases a SchemaContext is used and using a locater here would be bad because it would load the schema every time the method was called instead of using a version that was already loaded

File: docs/changehistory/NextVersion.md [hunk 3]
```diff
@@ -306,14 +325,37 @@ In addition to upgrading iTwin.js core dependencies to `4.0`, there are some oth
...
+
+- Added `SchemaInfo` interface with schema keys for a schema and it's references.  `Schema` implicitly supports this interface.
+- Some beta components had breaking changes and were moved to internal:
+  - `SchemaGraph`
+    - Now supports working with a `SchemaInfo` and a `SchemaContext` necessitating the init be made async.
+  - `SchemaMap`
+    - Use `Array<Schema>` in it's place.
+  - `SchemaCache`
+    - Updated to support caching partially loaded schemas, use `SchemaContext` to cache schemas in it's place.
+- Added helper method `SchemaFileUtility.writeSchemaToXmlString` to write schema xml to a string ``
```
Comment (grigasp) [inline docs/changehistory/NextVersion.md:360]:
```suggestion
- Added helper method `SchemaFileUtility.writeSchemaToXmlString` to write schema xml to a string
```

File: presentation/common/src/presentation-common/KoqPropertyValueFormatter.ts (hunk 2 above)
Comment (grigasp) [reply presentation/common/src/presentation-common/KoqPropertyValueFormatter.ts:62]: I see what you mean, but from API standpoint before your changes, I don't see why these methods couldn't have taken `ISchemaLocater`. All these functions care is to get a schema - they don't need to worry about where it comes from (fresh or from cache) - it's up to the caller to make sure they pass an `ISchemaLocater` that doesn't load schemas unnecessarily. And as far as I can tell that was **almost** happening except for passing the third argument to `getSchema`. However, due to what I believe is an API flaw, it was unnecessary.

In my opinion this confusion is caused by `SchemaContext` saying that it implements `ISchemaLocater` when it fact it doesn't. I think with changes in this PR it's even more apparent - `ISchemaLocater.getSchema` takes a required `SchemaContext` as the third argument and uses it to find referenced schemas. `SchemaContext.getSchema` doesn't use the third argument and instead looks for referenced schemas in its own cache. 

Consider this piece of code:
```ts
const context = new SchemaContext();
// add some schemas to `context`
const locater: ISchemaLocater = new SchemaContext();
const schemas = await locater.getSchema(key, matchType, context);
// referenced schemas in `context` won't be found
```
`getSchema` docs say that the third argument is going to be used for loading schema references, but it won't in this case.

To summarize, I think:
1. The change you made to our `getKoq` and `getFormattingProps` is okay.
2. `SchemaContext` should stop saying it implements `ISchemaLocater`.

**Review by grigasp — CHANGES_REQUESTED** [review]

File: presentation/common/src/presentation-common/KoqPropertyValueFormatter.ts (hunk 2 above)
Comment (grigasp) [reply presentation/common/src/presentation-common/KoqPropertyValueFormatter.ts:62]: I see what you mean, but from API standpoint before your changes, I don't see why these methods couldn't have taken `ISchemaLocater`. All these functions care is to get a schema - they don't need to worry about where it comes from (fresh or from cache) - it's up to the caller to make sure they pass an `ISchemaLocater` that doesn't load schemas unnecessarily. And as far as I can tell that was **almost** happening except for passing the third argument to `getSchema`. However, due to what I believe is an API flaw, it was unnecessary.

In my opinion this confusion is caused by `SchemaContext` saying that it implements `ISchemaLocater` when in fact it doesn't. I think with changes in this PR it's even more apparent - `ISchemaLocater.getSchema` takes a required `SchemaContext` as the third argument and says it's going to use it to find referenced schemas. `SchemaContext.getSchema` doesn't use the third argument and instead looks for referenced schemas in its own cache. 

Consider this piece of code:
```ts
const context = new SchemaContext();
// add some schemas to `context`
const locater: ISchemaLocater = new SchemaContext();
const schemas = await locater.getSchema(key, matchType, context);
// referenced schemas in `context` won't be found
```
`getSchema` docs say that the third argument is going to be used for loading schema references, but it won't in this case.

To summarize, I think:
1. The change you made to our `getKoq` and `getFormattingProps` is okay.
2. `SchemaContext` should stop saying it implements `ISchemaLocater`.

File: core/ecschema-metadata/src/Metadata/Schema.ts [hunk 4]
```diff
@@ -567,6 +567,9 @@ export class Schema implements CustomAttributeContainerProps {
     return schemaXml;
   }
 
+  /**
+   * Loads the schema header (name, version alias, label an description) from the input SchemaProps
```
Comment (calebmshafer) [inline core/ecschema-metadata/src/Metadata/Schema.ts:571]:
```suggestion
   * Loads the schema header (name, version alias, label and description) from the input SchemaProps
```

File: core/ecschema-metadata/src/Metadata/Schema.ts [hunk 5]
```diff
@@ -595,10 +598,25 @@ export class Schema implements CustomAttributeContainerProps {
       this._description = schemaProps.description;
   }
 
+  /**
+   * Loads the schema header (name, version alias, label an description) from the input SchemaProps
```
Comment (calebmshafer) [inline core/ecschema-metadata/src/Metadata/Schema.ts:602]:
```suggestion
   * Loads the schema header (name, version alias, label and description) from the input SchemaProps
```

File: presentation/common/src/presentation-common/KoqPropertyValueFormatter.ts (hunk 2 above)
Comment (ColinKerr) [reply presentation/common/src/presentation-common/KoqPropertyValueFormatter.ts:62]: That is a good point, let me review why SchemaContext implements ISchemaLocater and see if I can remove it.

Comment (ColinKerr) [reply presentation/common/src/presentation-common/KoqPropertyValueFormatter.ts:62]: I removed ISchemaLocater from SchemaContext and updated docs of ISchemaLocater to make its purpose more clear

File: common/changes/@itwin/ecschema-editing/cgk-async-schema-test_2023-05-10-22-29.json [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@itwin/ecschema-editing",
+      "comment": "Updated to use modified SchemaGraph",
```
[the remark this answers is not in the document]
Comment (ColinKerr) [reply common/changes/@itwin/ecschema-editing/cgk-async-schema-test_2023-05-10-22-29.json]: I changed the comment to remove the reference to schema graph

**Review by grigasp — APPROVED** [review]

**Comment by pmconne** [comment]
> I did both

You did?
image

File: docs/changehistory/NextVersion.md [hunk 7]
```diff
@@ -157,6 +161,21 @@ for await (const row of iModel.createQueryReader("SELECT * FROM bis.Element")) {
...
+- The interface `UnitConversion` has been renamed to [UnitConversionProps]($quantity).
+
+#### @itwin/ecschema-metadata
+
+- Renamed Interfaces
+  - The interface `FormatProps` has been changed to a type alias [SchemaItemFormatProps]($ecschema-metadata).
+  - The interface `UnitProps` has been renamed to [SchemaItemUnitProps]($ecschema-metadata).
+- `ISchemaLocater`
+  - `getSchema` and `getSchemaSync` methods now take a `Readonly<SchemaKey>` instead of a `SchemaKey` and the `SchemaContext` parameter is required
+  - Added `getSchemaInfo` method to facilitate asynchronous schema loading
```
Comment (pmconne) [inline docs/changehistory/NextVersion.md:177]: - The `FormatProps` interface has been replaced with the [SchemaItemFormatProps]($ecschema-metadata) type alias.
- The `UnitProps` interface has been renamed to [SchemaItemUnitProps]($ecschema-metadata).
- [ISchemaLocater.getSchema]($ecschema-metadata) and [ISchemaLocater.getSchemaSync]($ecschema-metadata) now take a `Readonly<SchemaKey>` instead of a [SchemaKey]($ecschema-metadata) and the [SchemaContext]($ecschema-metadata) parameter is no longer optional.
- Added [ISchemaLocater.getSchemaInfo]($ecschema-metadata) to facilitate asynchronous schema loading.

File: docs/changehistory/NextVersion.md [hunk 8]
```diff
@@ -306,14 +325,37 @@ In addition to upgrading iTwin.js core dependencies to `4.0`, there are some oth
...
 
-- The interface 'UnitProps' has been renamed to [SchemaItemUnitProps]($ecschema-metadata).
+### Other minor API changes
+
+- Added `SchemaInfo` interface with schema keys for a schema and it's references.  `Schema` implicitly supports this interface.
+- Some beta components had breaking changes and were moved to internal:
+  - `SchemaGraph`
+    - Now supports working with a `SchemaInfo` and a `SchemaContext` necessitating the init be made async.
+  - `SchemaMap`
+    - Use `Array<Schema>` in it's place.
```
Comment (pmconne) [inline docs/changehistory/NextVersion.md:357]:
```suggestion
    - Use `Array<Schema>` in its place.
```

File: docs/changehistory/NextVersion.md (hunk 7 above)
Comment (pmconne) [inline docs/changehistory/NextVersion.md:177]:
```suggestion
- The `FormatProps` interface has been replaced with the [SchemaItemFormatProps]($ecschema-metadata) type alias.
- The `UnitProps` interface has been renamed to [SchemaItemUnitProps]($ecschema-metadata).
- [ISchemaLocater.getSchema]($ecschema-metadata) and [ISchemaLocater.getSchemaSync]($ecschema-metadata) now take a `Readonly<SchemaKey>` instead of a [SchemaKey]($ecschema-metadata) and the [SchemaContext]($ecschema-metadata) parameter is no longer optional.
```

**Review by pmconne — APPROVED** [review]

**Review by calebmshafer — APPROVED** [review]
LGTM. Definitely need to work out some of the other inconsistencies in the APIs in addition to the ones that @grigasp mentioned but looks good.
